---
title: "miMeta Tutorial for Melody"
author: 
- name: "Zhoujingpeng Wei"
  affiliation: 
  - Department of Biostatistics and Medical Informatics, University of Wisconsin-Madison
  email: zwei74@wisc.edu
package: miMeta
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    code_folding: show
vignette: >
  %\VignetteIndexEntry{miMeta}
  %\VignetteEncoding{UTF-8}  
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

# Introduction

`miMeta` is a R package and it implements the new methods for meta-analysis of microbiome association studies that respect the unique features of microbiome data such as compositionality. This tutorial gives examples of applying a meta-analysis method called Melody [1] for microbial signature selection. Melody first generates summary statistics for each study and then uses the summary statistics across studies to select microbial signatures.

# Installation

Install package from GitHub.

```{r getPackage, echo=TRUE}
if(!require("miMeta", quietly = TRUE)){
  devtools::install_github("ZjpWei/miMeta")
}
```

Load the packages. 

```{r load, echo=TRUE, message=FALSE, warning=FALSE}
library("miMeta")
library("tidyverse")
```

# Meta-analysis given all samples of all participating studies

There are two approaches miMeta can run meta-analysis. 

Approach #1: If the individual-level sample data across all participating studies are available, the function `melody` can take these data as input to generate, harmonize, and combine summary association statistics across studies for accurate and robust identification of microbial signatures. 

Approach #2: If individual-level samples of each study are not in a central place, summary statistics can be generated for each study separately using the functions `melody.null.model` and `melody.get.summary`. The summary statistics can be transported to a central place to be harmonized and combined for signature selection using the function `melody.meta.summary`.

We demonstrate Approach #1 in this session and Approach #2 in the next session.

Here we use the datasets from two metagenomics studies of colorectal cancer (CRC) [2] to demonstrate the use of each function. The `CRC_abd` is a list of sample-by-feature matrices of relative abundance counts of 267 species under order *Clostridiales* from the two studies. The `CRC_meta` is a data frame including the sample-level variables from these two studies. In particular, the following variables are in the `CRC_meta` data:

* Sample identity: "Sample_ID"

* Study name: "Study"

* Disease status: "Group"

```{r echo=TRUE}
data("CRC_data")
CRC_abd <- CRC_data$CRC_abd
CRC_meta <- CRC_data$CRC_meta
```


```{r echo=TRUE, message=TRUE, warning=FALSE}
# Prepare input data
rel.abd <- list()
covariate.interest <- list()
for(d in c("FR-CRC", "DE-CRC")){
  rel.abd[[d]] <- CRC_abd[CRC_meta$Sample_ID[CRC_meta$Study == d],]
  disease <- as.numeric(CRC_meta$Group[CRC_meta$Study == d] == "CRC")
  names(disease) <- CRC_meta$Sample_ID[CRC_meta$Study == d]
  covariate.interest[[d]] <- data.frame(disease = disease)
}

meta.result <- melody(rel.abd = rel.abd, covariate.interest = covariate.interest, 
                      ref = "Coprococcus catus [ref_mOTU_v2_4874]", verbose = TRUE)
```

The first figure above shows the number of features shared among studies (here, all 266 species are present in both studies), and the second figure shows the meta-coefficient estimates for the selected microbial signatures for CRC.  

The following shows 20 coefficient estimates of the microbial features in the best model (the features with non-zero coefficients are the selected signatures):

```{r}
head(data.frame(coef = meta.result$disease$coef), n = 20)
```

By default, Melody only outputs the information about the best model. If users want to see the information about all models on the search path of subset size, set output.best.one = FALSE. See help(melody) for more details.

# Meta-analysis given samples of one study at a time
If the datasets of different studies live in different locations and cannot be conveniently shared among studies, we can first generate summary statistics for each study:

```{r echo=TRUE, message=FALSE, warning=FALSE}
# Generate summary statistics for each study
null.obj.FR <- melody.null.model(rel.abd = rel.abd["FR-CRC"], ref = "Coprococcus catus [ref_mOTU_v2_4874]")
summary.stats.FR <- melody.get.summary(null.obj = null.obj.FR,
                                       covariate.interest = covariate.interest["FR-CRC"])

null.obj.DE <- melody.null.model(rel.abd = rel.abd["DE-CRC"], ref = "Coprococcus catus [ref_mOTU_v2_4874]")
summary.stats.DE <- melody.get.summary(null.obj = null.obj.DE,
                                       covariate.interest = covariate.interest["DE-CRC"])
```

These summary statistics can be transported to a central location for meta-analysis:

```{r echo=TRUE, message=FALSE, warning=FALSE}
# Concatenate summary statistics
summary.stats.all <- c(summary.stats.FR, summary.stats.DE)
```


```{r echo=TRUE}
# Meta-analysis to harmonize and combine summary statistics across studies
meta.result.2 <- melody.meta.summary(summary.stats = summary.stats.all, verbose = TRUE)
```

The meta-analysis results generated this way are identical to those generated in the previous session.

# Meta-analysis of large-scale association scan
Microbiome association studies can involve a large number of covariates of interest (e.g., omics variables). We show here how to  use Melody to meta-analyze eight microbiome-metabolome association studies [3]. In these eight studies, we have 101 genera and 450 metabolites and we are interested in identifying genera that are associated with individual metabolites. This analysis takes approximately 20 minutes.

```{r echo=TRUE, message=FALSE, warning=FALSE}
# Load metabolite data: 
# You may see the following website on how to directly load data from 
# github into R https://github.com/ZjpWei/Melody/raw/main/Metabolite.rda
load(file = url("https://github.com/ZjpWei/Melody/raw/main/Metabolite.rda"))

# Change genura names
for(d in names(otu_data_lst)){
  colnames(otu_data_lst[[d]]) <-  gsub(".*;g__", "", colnames(otu_data_lst[[d]]))
}

# Get null model
null.obj <- melody.null.model(rel.abd = otu_data_lst, covariate.adjust = covariates_adjust_lst)

# Get summary statistics
summary.stats <- melody.get.summary(null.obj = null.obj, covariate.interest = cmpd_data_lst, cluster = cluster_data_lst)

# Meta-analysis
meta.scan.result <- melody.meta.summary(summary.stats = summary.stats, verbose = TRUE)
```

The following shows 20 coefficient estimates of the microbial features for 5 metabolites.

```{r echo=TRUE}
selected.num <- sort(unlist(lapply(meta.scan.result, function(d){sum(d$coef!=0)})), decreasing = TRUE)
top.cov.name <- names(selected.num)[1:min(5, length(selected.num))]
coef_mat <- sapply(meta.scan.result[top.cov.name], function(d){d$coef}, simplify = TRUE)
rownames(coef_mat) <- gsub(".*;g__","g__",rownames(coef_mat))
head(coef_mat, n = 20)
```

# Session information

```{r}
sessionInfo()
```

# Reference
1. Wei Z, Chen G, Tang ZZ. Melody: Meta-analysis of Microbiome Association Studies for Discovering Generalizable Microbial Signatures. Submitted.

2. Wirbel, Jakob et al. Meta-analysis of fecal metagenomes reveals global microbial signatures that are specific for colorectal cancer. 

3. Muller, E., Algavi, Y.M. & Borenstein, E. The gut microbiome-metabolome dataset collection: a curated resource for integrative meta-analysis. npj Biofilms Microbiomes 8, 79 (2022).
