---
title: "miMeta Tutorial"
author: 
- name: "Zhoujingpeng Wei"
  affiliation: 
  - Department of Biostatistics and Medical Informatics, University of Wisconsin-Madison
  email: zwei74@wisc.edu
package: miMeta
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    code_folding: show
vignette: >
  %\VignetteIndexEntry{miMeta}
  %\VignetteEncoding{UTF-8}  
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

# Introduction

`miMeta` is a R package and it implements the new methods for meta-analysis of microbiome association studies that respect the unique features of microbiome data such as compositionality. It first generate summary statistics (microbiome-disease association coefficient estimates and their variances) for individual studies. And then it combines the summary statistics across studies to select disease-associated microbial signatures based on the average absolute-abundance association coefficients inferred from the summary statistics. In particular, the selection of signature is operated through a best-subset selection.

# Installation

Download package from github.

```{r getPackage, echo=TRUE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("miMeta")
```

Load the package. 

```{r load, echo=TRUE}
library(miMeta)
library(tidyverse)
library(DT)
library(ggplot2)
```

# Meta-analysis given all samples of all participating studies

There are two approaches miMeta can run meta-analysis. Approach #1: If the individual-level sample data across all participating studies are available, the function melody can take these data as input, generate summary statistics for each study internally, and combine the summary statistics to select microbial signatures. Approach #2: If individual-level samples of each study are not in a central place and/or come in sequentially, summary statistics can be generated for each study separately using the function melody.get.summary. Then, the summary statistics of all participating studies can be combined using the function melody.merge.summary, and finally, meta-analyzed using the function melody.meta.summary. We demonstrate Approach #1 in this session and Approach #2 in the next session.

Here we use the datasets from the five metagenomics studies of colorectal cancer (CRC) [2] to demonstrate the use of each function. The "CRC_abd" is a feature-by-sample matrix of species-level relative abundance counts from the five studies. For the easy of demonstration, we include 267 species under order *Clostridiales*.  The "meta" is a data frame including the sample-level variables from the five studies. In particular, the following variables are in the meta data:

* Sample identity: "Sample_ID"

* Study name: "Study"

* Disease status: "Group"

```{r echo=TRUE}
data("CRC_abd", "meta", package = "miMeta")
table(meta[,c("Study", "Group")])
```

By default, melody only output the information about the best model selected. If users want to see the information about all models of different sizes on the search path, set output.best.one=FALSE. See help(melody) for more details.

The following is the coefficient of all the features in the best model (the features with non-zero coefficients are the selected signatures):

meta.result$coef

If verbose=TRUE, generate two plots shown below and print information about the progress of meta-analysis. 

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.cap="a plot for microbial feature overlap among studies: this plot shows the number of features shared among studies. "}
# Pick a reference taxa
ref <- "Coprococcus catus [ref_mOTU_v2_4874]"

# melody can analyze single study or multiple studies. 
meta.result <- melody(rel.abd = CRC_abd,
                      sample.data = meta,
                      sample.id = "Sample_ID",
                      study = "Study",
                      disease = "Group",
                      ref = ref,
                      verbose = TRUE)
```

```{r echo=FALSE, fig.cap="a plot showing the absolute-abundance coefficient estimates of the selected microbial features in the best model under the best subset size."}
taxa_tab <- data.frame(taxa = names(which(meta.result$coef!=0)),
                       coef = as.numeric(meta.result$coef[meta.result$coef!=0]))

taxa_tab %>% arrange(coef) %>% 
  mutate(taxa = factor(taxa, levels = taxa)) %>% 
  ggplot() + geom_point(aes(x= factor(taxa), y= coef)) + 
  theme_classic() + coord_flip() + ylab("coef") + 
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_blank(),
        axis.title.x = element_text(size = 10),
        panel.border = element_rect(colour = "black", fill=NA),
        legend.position = "right") +
  scale_x_discrete(position='bottom') +
  scale_fill_manual(values=c('lightgrey', 'darkgrey'), guide="none") +
  geom_hline(aes(yintercept = 0),colour="#990000", linetype="dashed")

```

# Meta-analysis given samples of one study at a time

In this session, we demonstrate how to perform meta-analysis given samples of one study at a time. We separate "CRC_abd" and "meta" by study to prepare datasets for the demonstration:

If the datasets of different studies live in different locations and cannot be conveniently transferred, we can first generate summary statistics for each study:

```{r echo=TRUE, message=FALSE, warning=FALSE}
# Geneeate summary statistisc for each study
meta_FR <- meta[meta$Study == "FR-CRC",]
summary.FR <- melody.get.summary(rel.abd = CRC_abd[,meta_FR$Sample_ID],
                                 sample.data = meta_FR,
                                 sample.id = "Sample_ID",
                                 study = "Study",
                                 disease = "Group",
                                 ref = ref)

meta_DE <- meta[meta$Study == "DE-CRC",]
summary.DE <- melody.get.summary(rel.abd = CRC_abd[,meta_DE$Sample_ID],
                                 sample.data = meta_DE,
                                 sample.id = "Sample_ID",
                                 study = "Study",
                                 disease = "Group",
                                 ref = ref)

meta_CN <- meta[meta$Study == "CN-CRC",]
summary.CN <- melody.get.summary(rel.abd = CRC_abd[,meta_CN$Sample_ID],
                                 sample.data = meta_CN,
                                 sample.id = "Sample_ID",
                                 study = "Study",
                                 disease = "Group",
                                 ref = ref)

meta_US <- meta[meta$Study == "US-CRC",]
summary.US <- melody.get.summary(rel.abd = CRC_abd[,meta_US$Sample_ID],
                                 sample.data = meta_US,
                                 sample.id = "Sample_ID",
                                 study = "Study",
                                 disease = "Group",
                                 ref = ref)

meta_AT <- meta[meta$Study == "AT-CRC",]
summary.AT <- melody.get.summary(rel.abd = CRC_abd[,meta_AT$Sample_ID],
                                 sample.data = meta_AT,
                                 sample.id = "Sample_ID",
                                 study = "Study",
                                 disease = "Group",
                                 ref = ref)
```

The summary data can be sent to a central place and get merged:

```{r echo=TRUE, message=FALSE, warning=FALSE}
# Merge summary statistics
summary.merge <- melody.merge.summary(list(summary.FR, summary.DE, summary.CN, summary.US, summary.AT))
```

The merged object can be used to run meta-analysis:

```{r echo=TRUE}
# Meta-analysis for merged summary statistics
meta.result.merge <- melody.meta.summary(Melody = summary.merge, verbose = TRUE)
```

The meta-analysis results generated in this way are identical to the results generated in the previous session.

```{r echo=FALSE, fig.cap="a plot showing the absolute-abundance coefficient estimates of the selected microbial features in the best model under the best subset size by using merged summary statistics."}
taxa_tab <- data.frame(taxa = names(which(meta.result$coef!=0)),
                       coef = as.numeric(meta.result$coef[meta.result$coef!=0]))

taxa_tab %>% arrange(coef) %>% 
  mutate(taxa = factor(taxa, levels = taxa)) %>% 
  ggplot() + geom_point(aes(x= factor(taxa), y= coef)) + 
  theme_classic() + coord_flip() + ylab("coef") + 
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_blank(),
        axis.title.x = element_text(size = 10),
        panel.border = element_rect(colour = "black", fill=NA),
        legend.position = "right") +
  scale_x_discrete(position='bottom') +
  scale_fill_manual(values=c('lightgrey', 'darkgrey'), guide="none") +
  geom_hline(aes(yintercept = 0),colour="#990000", linetype="dashed")

```
# Session information

```{r}
sessionInfo()
```

# Reference
1. Wei Z, Chen G, Tang ZZ. Melody identifies generalizable microbial signatures in microbiome association meta-analysis. Submitted.

2. Wirbel, Jakob et al. Meta-analysis of fecal metagenomes reveals global microbial signatures that are specific for colorectal cancer. 
